### PR TITLE
Open Write

### PR DESCRIPTION
**Pull request by seanmcc-msft**
I need some of James' code that only exists in the 73 base branch.  Continued from https://github.com/Azure/azure-sdk-for-net/pull/12662

### CODE REVIEW

**Comment by PaulVrugt** [comment]
Just here to cheer you on!

**Comment by seanmcc-msft** [comment]
The main differences between this and https://github.com/Azure/azure-sdk-for-net/pull/12662 is that now I support buffer sizes > max int, and all writes to the service will be the same length (buffer size).

**Comment by seanmcc-msft** [comment]
@kasobol-msft, @jaschrep-msft

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs [hunk 1]
```diff
@@ -54,13 +54,24 @@ private class BufferPartition
         /// </summary>
         private List<BufferPartition> BufferSet { get; } = new List<BufferPartition>();
 
+        /// <summary>
+        /// Size to rent from MemoryPool during Writes.
+        /// </summary>
+        public int ArrayPoolRentalSize { get; }
```
Comment (jaschrep-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:60]: This appears to serve the same purpose as `MaxArraySize`. What is the difference?

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:60]: Nothing I guess, I didn't understand what `MaxArraySize` was from the description.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:60]: I'll remove in the next iteration.

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs [hunk 2]
```diff
@@ -315,16 +326,44 @@ public override void SetLength(long value)
...
+                }
+
+                int copied = Math.Min(currentBuffer.Buffer.Length - currentBuffer.DataLength, count);
+                Array.Copy(buffer, offset, currentBuffer.Buffer, currentBuffer.DataLength, copied);
+                currentBuffer.DataLength += copied;
+                count -= copied;
+                offset += copied;
+                Position += copied;
+            }
+            while (count > 0);
```
Comment (jaschrep-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:351]: Nit: if someone writes logic that calls this method with count = 0 as part of their own loop, I'd rather have caught that at the start and avoided all these operations in exchange for one extra check at the start of each write.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:351]: I'll flip it to a normal while loop.

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs [hunk 3]
```diff
@@ -54,13 +54,24 @@ private class BufferPartition
...
+        public int ArrayPoolRentalSize { get; }
+
         private PooledMemoryStream(ArrayPool<byte> arrayPool, long absolutePosition, int maxArraySize)
         {
             AbsolutePosition = absolutePosition;
             ArrayPool = arrayPool;
             MaxArraySize = maxArraySize;
         }
 
+        public PooledMemoryStream(int arrayPoolRentalSize)
```
Comment (jaschrep-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:69]: If we consolidate `MaxArraySize` and `ArrayPoolRentalSize` then maybe we should consolidate these constructors. `arrayPool` and `absolutePosition` could be default parameters.

**Comment by jaschrep-msft** [comment]
Please add tests for your new `PooledMemoryStream` functionality to `Azure.Storage.Common/tests/PooledMemoryStreamTests.cs`

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs (hunk 1 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:60]: Thanks. Feel free to refactor the name/docstring to be more clear, as I do see the confusion.

**Comment by seanmcc-msft** [comment]
> Please add tests for your new `PooledMemoryStream` functionality to `Azure.Storage.Common/tests/PooledMemoryStreamTests.cs`

Will do.  I want to make sure One SDK will let us ship this first.

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:69]: Will do.

**Review by kasobol-msft — COMMENTED** [review]
LGTM. We should add decent test coverage and make sure we're hitting all these conditions.

**Comment by seanmcc-msft** [comment]
> LGTM. We should add decent test coverage and make sure we're hitting all these conditions.

I'll add a ton of tests, but I want to make sure One SDK will let us do this first.

**Review by jaschrep-msft — COMMENTED** [review]
Can OpenWrite not begin writing to the middle of a blob (when the blob type allows) or is it only append and overwrite?

Overall structure looks good, I'll take a closer look once we get design approval and all that stuff in place.

I've been thinking about `PooledMemoryStream` and how we're now using it in a more general purpose way, whereas it was previously built solely for `PartitionedUploader`. The property `AbsolutePosition` Doesn't make sense outside of `PartitionedUploader` and the internal static methods that read in a section of another stream are also very clearly situational to the uploader. With some of the other parallel work going on right now, I don't think it can be addressed in this PR. I think we need to accept this slightly bloated class for now. I filed #12820 to address this in the future.

**Comment by seanmcc-msft** [comment]
> Can OpenWrite not begin writing to the middle of a blob (when the blob type allows) or is it only append and overwrite?

This is an interesting question.  For Block Blobs, Append Blobs, and Data Lake Files, it's not possible to overwrite the middle of a blob/file.  I believe it is possible with SMB Files and Page Blobs.''

In track 1, it doesn't appear we allowed users to begin Open Write in the middle of a Page Blob - https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Blob/CloudPageBlob.cs#L308, and Data Lake Files only existed in preview.

My vote would be to only support Open Write for the overwrite and append cases.

**Comment by PaulVrugt** [comment]
The openwrite in V11 supported seeking to the middle of the file for page blobs. This might come in handy for our particular use case. I think it's worth considering to continue this support in the V12 version

**Comment by seanmcc-msft** [comment]
> The openwrite in V11 supported seeking to the middle of the file for page blobs. This might come in handy for our particular use case. I think it's worth considering to continue this support in the V12 version

I think we're considering hold off on Seek() for the first (preview) version of OpenWrite(), but we could added it for Page Bobs and SMB Files in the second version.  There's some complexity around how to handle data that has been buffered by not yet flushed when the user calls Seek().

**Comment by PaulVrugt** [comment]
> > The openwrite in V11 supported seeking to the middle of the file for page blobs. This might come in handy for our particular use case. I think it's worth considering to continue this support in the V12 version
> 
> I think we're considering hold off on Seek() for the first (preview) version of OpenWrite(), but we could added it for Page Bobs and SMB Files in the second version. There's some complexity around how to handle data that has been buffered by not yet flushed when the user calls Seek().

I think I can live with that 😊

File: sdk/storage/Azure.Storage.Common/src/Shared/WriteStream.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+                    offset,
+                    remainingSpace,
+                    cancellationToken).ConfigureAwait(false);
+                remaining -= remainingSpace;
+                offset += remainingSpace;
+
+                // Upload bytes.
+                await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+                while (remaining > 0)
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/WriteStream.cs:109]: Added a comment.

File: sdk/storage/Azure.Storage.Common/src/Shared/WriteStream.cs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                count,
+                cancellationToken)
+            .ConfigureAwait(false);
+
+        protected virtual async Task WriteInternal(
+            bool async,
+            byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/WriteStream.cs:83]: Fixed.

File: sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs [hunk 6]
```diff
@@ -1341,6 +1341,147 @@ public new AppendBlobClient WithVersion(string versionId)
...
+        {
+            DiagnosticScope scope = ClientDiagnostics.CreateScope($"{nameof(AppendBlobClient)}.{nameof(OpenWrite)}");
+
+            try
+            {
+                scope.Start();
+
+                long position = 0;
+
+                if (options != null && options.Overwrite)
```
Comment (jaschrep-msft) [inline sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:1435]: Nit: `options?.Overwrite` since this just looks like safe access.

**Comment by PaulVrugt** [comment]
Well that's disappointing...

**Comment by amishra-dev** [comment]
> Well that's disappointing...

@PaulVrugt we are starting a new PR as the base branch needs to change.

**Comment by seanmcc-msft** [comment]
@PaulVrugt, this was closed automatically because I merged the base branch.  I'll re-open a new PR was the master branch as base.  We are still planning to do Open Read and Open Write, sorry for the confusion.

**Comment by seanmcc-msft** [comment]
> > Please add tests for your new `PooledMemoryStream` functionality to `Azure.Storage.Common/tests/PooledMemoryStreamTests.cs`
> 
> Will do. I want to make sure One SDK will let us ship this first.

Done in next iteration.

**Pull request closed without merging**